std/experimental/allocator/building_blocks/region.d

```diff
@@ -878,3 +890,385 @@ version(Posix) @system unittest
 }
+
+/**
+The threadsafe version of the `Region` allocator.
+Allocations and deallocations are lock-free based using the `cas`
+functionality from `core.atomic`.
+*/
+shared struct SharedRegion(ParentAllocator = NullAllocator,
+    uint minAlign = platformAlignment,
+    Flag!"growDownwards" growDownwards = No.growDownwards)
+{
+    static assert(minAlign.isGoodStaticAlignment);
+    static assert(ParentAllocator.alignment >= minAlign);
+
+    import std.traits : hasMember;
+    import std.typecons : Ternary;
+
+    // state
+    /**
+    The _parent allocator. Depending on whether $(D ParentAllocator) holds state
+    or not, this is a member variable or an alias for
+    `ParentAllocator.instance`.
+    */
+    static if (stateSize!ParentAllocator)
+    {
+        ParentAllocator parent;
+    }
+    else
+    {
+        alias parent = ParentAllocator.instance;
+    }
+    private shared void* _current, _begin, _end;
+
+    /**
+    Constructs a region backed by a user-provided store. Assumes $(D store) is
+    aligned at $(D minAlign). Also assumes the memory was allocated with $(D
+    ParentAllocator) (if different from $(D NullAllocator)).
+
+    Params:
+    store = User-provided store backing up the region. $(D store) must be
+    aligned at $(D minAlign) (enforced with $(D assert)). If $(D
+    ParentAllocator) is different from $(D NullAllocator), memory is assumed to
+    have been allocated with $(D ParentAllocator).
+    n = Bytes to allocate using $(D ParentAllocator). This constructor is only
+    defined If $(D ParentAllocator) is different from $(D NullAllocator). If
+    $(D parent.allocate(n)) returns $(D null), the region will be initialized
+    as empty (correctly initialized but unable to allocate).
+    */
+    this(ubyte[] store)
+    {
+        store = cast(ubyte[])(store.roundUpToAlignment(alignment));
+        store = store[0 .. $.roundDownToAlignment(alignment)];
+        assert(store.ptr.alignedAt(minAlign));
+        assert(store.length % minAlign == 0);
+        _begin = cast(typeof(_begin)) store.ptr;
+        _end = cast(typeof(_end)) (store.ptr + store.length);
+        static if (growDownwards)
+            _current = _end;
+        else
+            _current = cast(typeof(_current)) store.ptr;
+    }
+
+    /// Ditto
+    static if (!is(ParentAllocator == NullAllocator))
+    this(size_t n)
+    {
+        this(cast(ubyte[])(parent.allocate(n.roundUpToAlignment(alignment))));
+    }
+
+    /**
+    Rounds the given size to a multiple of the `alignment`
+    */
+    size_t goodAllocSize(size_t n)
+    {
+        return n.roundUpToAlignment(alignment);
+    }
+
+    /**
+    Alignment offered.
+    */
+    alias alignment = minAlign;
+
+    /**
+    Allocates $(D n) bytes of memory. The allocation is served by atomically incrementing
+    a pointer which keeps track of the current used space.
+
+    To increase performance, allocations which exceed the region capacity could cause other
```

Comment:
Taking a second look at this, I was wrong in my comments.
The cas loop is useful because it prevents intermitent failures by checking bounds inside the loop.